lib/web/servers.go

```diff
@@ -157,3 +158,53 @@ func (h *Handler) getDesktopHandle(w http.ResponseWriter, r *http.Request, p htt
 	// to see the desktop once in the UI, so just take the first one.
 	return ui.MakeDesktop(windowsDesktops[0]), nil
 }
+
+// desktopIsActive checks if a desktop has an active session and if so, returns a one time token that is needed
+// to start a session to said desktop.
+//
+// GET /v1/webapi/sites/:site/desktops/:desktopName/active
+//
+// Response body:
+//
+// {"active": bool}
+func (h *Handler) desktopIsActive(w http.ResponseWriter, r *http.Request, p httprouter.Params, ctx *SessionContext, site reversetunnel.RemoteSite) (interface{}, error) {
+	desktopName := p.ByName("desktopName")
+	trackers, err := h.auth.proxyClient.GetActiveSessionTrackers(r.Context())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	checker, err := ctx.GetUserAccessChecker()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	for _, tracker := range trackers {
+		if tracker.GetSessionKind() == types.WindowsDesktopSessionKind && tracker.GetState() == types.SessionState_SessionStateRunning && tracker.GetDesktopName() == desktopName {
+			desktops, err := h.auth.accessPoint.GetWindowsDesktops(r.Context(), types.WindowsDesktopFilter{Name: tracker.GetDesktopName()})
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			if len(desktops) == 0 {
+				break
+			}
+
+			err = checker.CheckAccess(desktops[0], services.AccessMFAParams{})
```

Comment:
Is there any side effects of using the empty `services.AccessMFAParams{}` here ? I'm not too up to scratch on our access engine, but I'd worry that if a resource enforced MFA we'd reject them all as access denied.

Reply:
Took a look -- it comes down to this function which takes its cues as to whether to check for MFA credentials from `AccessMFAParams{}`. 

https://github.com/gravitational/teleport/blob/e7b2bd6326c5cb743fdb5d42a4c3e469b04d70c4/lib/services/role.go#L1988-L2128

If we pass the empty structure then it takes it to mean "MFA not required".

When another user has an active session and per session MFA is required, the user will first see 

<img width="1075" alt="image" src="https://user-images.githubusercontent.com/13578537/198389933-dbb562e8-9ff8-4d10-9ba4-793d925b2941.png">

and if they continue, they will be prompted for MFA

<img width="750" alt="image" src="https://user-images.githubusercontent.com/13578537/198390103-8b42d192-f965-4901-b58a-a13acb5b5cef.png">

Technically speaking, that first prompt can be taken to be leaking information to a user not authenticated to the Teleport administrator's standard: 

In the non-MFA case, if a user doesn't have permission to access to the desktop they trying to connect to, we just [return the default case](https://github.com/gravitational/teleport/blob/e7b2bd6326c5cb743fdb5d42a4c3e469b04d70c4/lib/web/servers.go#L203-L207) (`return desktopIsActive{false}, nil`), which means we don't leak information about whether somebody has an active session on that desktop (or whether a desktop with such a name even exists).

Whereas currently written, in the MFA case, we can leak information about whether a desktop exists and whether another user is using it to somebody who has access to that desktop according to RBAC, but who hasn't recently completed MFA, so is in effect unauthorized.

Fixing it would require a substantial rethink about how we do MFA with desktop access (currently its done through TDP, meaning we need to make the connection before doing it, which is exactly what we're preventing with the "another session is active" screen here). Pedantic, but ignoring the administrator's wishes and the inconsistent security behavior still leaves a bad taste in my mouth.

@strideynet thoughts?

Reply:
I'm more than happy to proceed with the current behaviour, I was more just interested in checking this had been considered. I'm not sure it leaks anything overly sensitive - and a substantial rethink may well be more engineering effort than this is worth.

Reply:
Agree